twins/nats/pubsub.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package nats
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/mainflux/mainflux"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/twins"
+	"github.com/mainflux/mainflux/twins/mqtt"
+	"github.com/mainflux/senml"
+	"github.com/nats-io/go-nats"
+)
+
+const (
+	queue = "twins"
+	input = "channel.>"
+)
+
+var mqttOp = map[string]string{
+	"stateSucc": "state/success",
+	"stateFail": "state/failure",
+}
+
+type pubsub struct {
+	natsClient *nats.Conn
+	mqttClient mqtt.Mqtt
+	logger     log.Logger
+	twins      twins.TwinRepository
+	states     twins.StateRepository
+}
+
+// Subscribe to appropriate NATS topic
+func Subscribe(nc *nats.Conn, mc mqtt.Mqtt, tr twins.TwinRepository, sr twins.StateRepository, logger log.Logger) {
+	ps := pubsub{
+		natsClient: nc,
+		mqttClient: mc,
+		logger:     logger,
+		twins:      tr,
+		states:     sr,
+	}
+	ps.natsClient.QueueSubscribe(input, queue, ps.handleMsg)
+}
+
+func (ps pubsub) handleMsg(m *nats.Msg) {
+	var msg mainflux.Message
+	err := proto.Unmarshal(m.Data, &msg)
+	if err != nil {
+		ps.logger.Warn(fmt.Sprintf("Unmarshalling failed: %s", err))
+		return
+	}
+	if msg.Channel == ps.mqttClient.Topic() {
+		return
+	}
+
+	var b []byte
+	var id string
+	defer ps.mqttClient.Publish(&id, &err, mqttOp["stateSucc"], mqttOp["stateFail"], &b)
+
+	tw, err := ps.twins.RetrieveByThing(context.TODO(), msg.Publisher)
+	if err != nil {
+		ps.logger.Warn(fmt.Sprintf("Retrieving twin for %s failed: %s", msg.Publisher, err))
+		return
+	}
+
+	var recs []senml.Record
+	if err := json.Unmarshal(msg.Payload, &recs); err != nil {
+		ps.logger.Warn(fmt.Sprintf("Unmarshal payload for %s failed: %s", msg.Publisher, err))
+		return
+	}
+
+	st, err := ps.states.RetrieveLast(context.TODO(), tw.ID)
+	if err != nil {
+		ps.logger.Warn(fmt.Sprintf("Retrieve last state for %s failed: %s", msg.Publisher, err))
+		return
+	}
+
+	if save := prepareState(&st, &tw, recs, msg); !save {
+		ps.logger.Info(fmt.Sprintf("No persistent attributes for %s for %s", msg.Subtopic, msg.Publisher))
+		return
+	}
+
+	if err := ps.states.Save(context.TODO(), st); err != nil {
+		ps.logger.Warn(fmt.Sprintf("Updating state for %s failed: %s", msg.Publisher, err))
+		return
+	}
```

Comment:
Is it possible to move this logic to Twin service? Also, can you try to simplify this logic? Retrieving twin and state, then storing state for each message is very heavy. You can consider replacing this with a single, more complex update query.

Reply:
We have to store the state for each message. That is the business logic of the service. Any single attribute change implies the new state. For now, for the purpose of simplicity, we send only one measurement per message. It's by design for the first version. 

Sure, I can put this in service. Is it a good idea? @nmarcetic ^

For retrieving twin and state - the last state must be retrieved, the twin also. Because we have to compare definition and measurement. Maybe some form of caching of twins and states? @nmarcetic 

@dusanb94 can you give me some pointers towards the simplification of this func. You should keep in mind that we keep twin and its related stated in the separate db collections.

Reply:
@darkodraskovic Maybe you can checkout MongoDB doc and [update method](https://docs.mongodb.com/manual/reference/operator/update/positional/), more complex query which will do the job is more welcome :) I agree with @dusanb94

Reply:
Yep, I agree about moving this into the service. Nats package should contain only nats logic.

Reply:
@nmarcetic and @dusanb94 I think you do not take into account that twin and state are in different databases.

Reply:
Moved the logic into the twin service.